File: android-native/app/src/main/kotlin/com/wordwheel/game/ui/TopBar.kt
package com.wordwheel.game.ui

import androidx.compose.animation.core.LinearOutSlowInEasing
import androidx.compose.animation.core.animateIntAsState
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wordwheel.game.theme.GameColors

/** Layered coin glyph. Two concentric circles + a centred star give a
 *  more "game-y" point-counter than the plain green disc that came
 *  before. Drawn entirely in Compose so we don't need a drawable
 *  resource and it scales cleanly on every density. */
@Composable
private fun CoinIcon(modifier: Modifier = Modifier, size: Int = 22) {
    Box(
        modifier = modifier
            .size(size.dp)
            .clip(CircleShape)
            .background(
                Brush.linearGradient(
                    colors = listOf(Color(0xFFFFE070), Color(0xFFFFAA00)),
                )
            )
            .border(width = 2.dp, color = Color(0xFF8B5A00), shape = CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = "★",
            color = Color(0xFF8B5A00),
            fontSize = (size * 0.6f).sp,
            fontWeight = FontWeight.Black,
        )
    }
}

@Composable
fun TopBar(coins: Int, found: Int, total: Int, level: Int, streak: Int = 0) {
    // Tween-animate the displayed coin count when it changes — the spin
    // reward (and word-completion +2) ticks up smoothly instead of
    // snapping. ~600ms is short enough to feel responsive but long
    // enough to read the change.
    val displayedCoins by animateIntAsState(
        targetValue = coins,
        animationSpec = tween(durationMillis = 600, easing = LinearOutSlowInEasing),
        label = "coinCounter",
    )
    // Track previous count so we can pulse the icon when the value
    // jumps by a meaningful amount (skip the +2 from a single word).
    var lastCoins by remember { mutableStateOf(coins) }
    var pulseActive by remember { mutableStateOf(false) }
    LaunchedEffect(coins) {
        if (coins - lastCoins >= 5) {
            pulseActive = true
            kotlinx.coroutines.delay(240)
            pulseActive = false
        }
        lastCoins = coins
    }
    val pulse by animateFloatAsState(
        targetValue = if (pulseActive) 1.25f else 1f,
        animationSpec = tween(durationMillis = 220),
        label = "coinPulse",
    )

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(22.dp))
            .background(GameColors.TopBarBg)
            .padding(horizontal = 14.dp, vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        CoinIcon(modifier = Modifier.scale(pulse), size = 22)
        Spacer(Modifier.width(8.dp))
        Text(
            text = displayedCoins.toString(),
            color = Color.White,
            fontSize = 17.sp,
            fontWeight = FontWeight.Bold,
        )

        Spacer(Modifier.width(14.dp))

        // Words badge
        Box(
            modifier = Modifier
                .clip(RoundedCornerShape(14.dp))
                .background(GameColors.BadgeBlue)
                .padding(horizontal = 10.dp, vertical = 4.dp),
        ) {
            Text(
                text = "W  $found/$total",
                color = Color.White,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
            )
        }

        // Streak badge — only shown once the player has a streak going.
        // Hidden at streak==0 to avoid noise on first install.
        if (streak > 0) {
            Spacer(Modifier.width(8.dp))
            Box(
                modifier = Modifier
                    .clip(RoundedCornerShape(14.dp))
                    .background(Color(0x40FF7028))
                    .padding(horizontal = 8.dp, vertical = 4.dp),
            ) {
                Text(
                    text = "🔥 $streak",
                    color = Color.White,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.SemiBold,
                )
            }
        }

        Spacer(Modifier.weight(1f))

        // Level badge
        Box(
            modifier = Modifier
                .clip(RoundedCornerShape(14.dp))
                .background(Color(0x28FFFFFF))
                .padding(horizontal = 12.dp, vertical = 4.dp),
        ) {
            Text(
                text = "Lv.$level",
                color = Color.White,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
            )
        }
    }
}

File: ios-native/WordWheel/Views/TopBar.swift
import SwiftUI

/// Layered coin glyph. Two concentric circles + a centred star give a
/// more "game-y" point-counter than the plain green disc that came
/// before. Drawn entirely in SwiftUI so it scales cleanly on every
/// density without an asset bundle.
private struct CoinIcon: View {
    let size: CGFloat

    var body: some View {
        ZStack {
            Circle()
                .fill(LinearGradient(
                    colors: [Color(hex: 0xFFFFE070), Color(hex: 0xFFFFAA00)],
                    startPoint: .topLeading, endPoint: .bottomTrailing,
                ))
                .overlay(Circle().stroke(Color(hex: 0xFF8B5A00), lineWidth: 2))
            Text("★")
                .font(.system(size: size * 0.6, weight: .black))
                .foregroundColor(Color(hex: 0xFF8B5A00))
        }
        .frame(width: size, height: size)
    }
}

struct TopBar: View {
    let coins: Int
    let found: Int
    let total: Int
    let level: Int
    var streak: Int = 0

    /// The displayed (animated) coin count. Tracks `coins` lazily so a
    /// jump from 200→225 visibly counts up instead of snapping.
    @State private var displayedCoins: Int = 0
    /// Set briefly when coins jump by a meaningful amount (>=5) so we
    /// can pulse the icon. Single +2 word rewards skip this to avoid
    /// constant scaling noise.
    @State private var pulseActive: Bool = false

    var body: some View {
        HStack(spacing: 0) {
            CoinIcon(size: 22)
                .scaleEffect(pulseActive ? 1.25 : 1.0)
                .animation(.easeOut(duration: 0.22), value: pulseActive)
            Spacer().frame(width: 8)
            Text("\(displayedCoins)")
                .font(.system(size: 17, weight: .bold))
                .foregroundColor(.white)
                .contentTransition(.numericText())

            Spacer().frame(width: 14)

            // Words badge
            Text("W  \(found)/\(total)")
                .font(.system(size: 14, weight: .semibold))
                .foregroundColor(.white)
                .padding(.horizontal, 10).padding(.vertical, 4)
                .background(RoundedRectangle(cornerRadius: 14).fill(GameColors.badgeBlue))

            // Streak badge — only when there's an active streak
            if streak > 0 {
                Spacer().frame(width: 8)
                Text("🔥 \(streak)")
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundColor(.white)
                    .padding(.horizontal, 8).padding(.vertical, 4)
                    .background(RoundedRectangle(cornerRadius: 14)
                        .fill(Color(red: 1, green: 0.44, blue: 0.16, opacity: 0.25)))
            }

            Spacer()

            // Level badge
            Text("Lv.\(level)")
                .font(.system(size: 14, weight: .semibold))
                .foregroundColor(.white)
                .padding(.horizontal, 12).padding(.vertical, 4)
                .background(RoundedRectangle(cornerRadius: 14).fill(Color.white.opacity(0.157)))
        }
        .padding(.horizontal, 14).padding(.vertical, 8)
        .background(RoundedRectangle(cornerRadius: 22).fill(GameColors.topBarBg))
        .onAppear { displayedCoins = coins }
        .onChange(of: coins) { newValue in
            let delta = newValue - displayedCoins
            // Count-up tween — ~600ms feels good for spin rewards while
            // small +2 word adds still settle quickly.
            withAnimation(.easeOut(duration: 0.6)) {
                displayedCoins = newValue
            }
            // Pulse only on chunky jumps (spin reward, level bonus).
            if delta >= 5 {
                pulseActive = true
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.24) {
                    pulseActive = false
                }
            }
        }
    }
}
